src/main/java/org/opensearch/neuralsearch/processor/rerank/QueryContextSourceFetcher.java

```diff
@@ -0,0 +1,103 @@
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.ObjectPath;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.neuralsearch.query.ext.RerankSearchExtBuilder;
+import org.opensearch.search.SearchExtBuilder;
+
+/**
+ * Context Source Fetcher that gets context from the rerank query ext.
+ */
+public class QueryContextSourceFetcher implements ContextSourceFetcher {
+
+    public static final String NAME = "query_context";
+    public static final String QUERY_TEXT_FIELD = "query_text";
+    public static final String QUERY_TEXT_PATH_FIELD = "query_text_path";
+
+    @Override
+    public void fetchContext(SearchRequest searchRequest, SearchResponse searchResponse, ActionListener<Map<String, Object>> listener) {
+        try {
+            List<SearchExtBuilder> exts = searchRequest.source().ext();
+            Map<String, Object> params = RerankSearchExtBuilder.fromExtBuilderList(exts).getParams();
+            Map<String, Object> scoringContext = new HashMap<>();
+            if (!params.containsKey(NAME)) {
+                throw new IllegalArgumentException(String.format(Locale.ROOT, "must specify %s", NAME));
+            }
+            Object ctxObj = params.remove(NAME);
+            if (!(ctxObj instanceof Map<?, ?>)) {
+                throw new IllegalArgumentException(String.format(Locale.ROOT, "%s must be a map", NAME));
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> ctxMap = (Map<String, Object>) ctxObj;
+            if (ctxMap.containsKey(QUERY_TEXT_FIELD)) {
+                if (ctxMap.containsKey(QUERY_TEXT_PATH_FIELD)) {
+                    throw new IllegalArgumentException(
+                        String.format(Locale.ROOT, "Cannot specify both \"%s\" and \"%s\"", QUERY_TEXT_FIELD, QUERY_TEXT_PATH_FIELD)
+                    );
+                }
+                scoringContext.put(QUERY_TEXT_FIELD, (String) ctxMap.get(QUERY_TEXT_FIELD));
+            } else if (ctxMap.containsKey(QUERY_TEXT_PATH_FIELD)) {
+                String path = (String) ctxMap.get(QUERY_TEXT_PATH_FIELD);
+                // Convert query to a map with io/xcontent shenanigans
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                XContentBuilder builder = XContentType.CBOR.contentBuilder(baos);
+                searchRequest.source().toXContent(builder, ToXContent.EMPTY_PARAMS);
+                builder.close();
+                ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+                XContentParser parser = XContentType.CBOR.xContent().createParser(NamedXContentRegistry.EMPTY, null, bais);
+                Map<String, Object> map = parser.map();
```

Comment:
there's this [searchRequestMap](https://github.com/opensearch-project/OpenSearch/blob/main/modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/SearchRequestMap.java) which does something similar but it doesn't expose all possible keys - which makes sense since the entire query DSL is rather large...